File: haskell/src/Todo/Adapters/MemoryRepo.hs
-- | In-memory repository.
--
-- A real implementation of the port that happens to be fast and disposable, and
-- it is verified by the same contract test as the SQLite one. A fake that is
-- never verified is just a second bug farm.
module Todo.Adapters.MemoryRepo
  ( newInMemoryRepository
  ) where

import Data.IORef (modifyIORef', newIORef, readIORef)
import Data.List (find)
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Text (Text)

import Todo.Domain.Model (Todo (..), isDone)
import Todo.Domain.Ports (TodoRepository (..))

newInMemoryRepository :: IO TodoRepository
newInMemoryRepository = do
  ref <- newIORef (Map.empty :: Map Text Todo)
  pure (fromRef ref)
  where
    fromRef ref = TodoRepository
      { repoAdd     = \todo -> modifyIORef' ref (Map.insert (todoId todo) todo)
      , repoReplace = \todo -> modifyIORef' ref (Map.insert (todoId todo) todo)
      , repoRemove  = \tid  -> modifyIORef' ref (Map.delete tid)
      , repoGet     = \tid  -> Map.lookup tid <$> readIORef ref
      , repoFindActiveByTitle = \title -> do
          todos <- readIORef ref
          pure (find (\t -> todoTitle t == title && not (isDone t)) (Map.elems todos))
      , repoAll = Map.elems <$> readIORef ref
      }

File: python/src/todo/adapters/memory_repo.py
"""In-memory repository.

This ships in `src/`, not in `tests/`, on purpose. It is a real implementation
of the port that happens to be fast and disposable, and it is verified by the
same contract test as the SQLite one. A fake that is never verified is just a
second bug farm.
"""

from __future__ import annotations

from ..domain.model import Todo
from ..domain.ports import TodoRepository


class InMemoryTodoRepository(TodoRepository):
    def __init__(self) -> None:
        self._todos: dict[str, Todo] = {}

    def add(self, todo: Todo) -> None:
        self._todos[todo.id] = todo

    def replace(self, todo: Todo) -> None:
        self._todos[todo.id] = todo

    def remove(self, todo_id: str) -> None:
        self._todos.pop(todo_id, None)

    def get(self, todo_id: str) -> Todo | None:
        return self._todos.get(todo_id)

    def find_active_by_title(self, title: str) -> Todo | None:
        for todo in self._todos.values():
            if todo.title == title and not todo.is_done:
                return todo
        return None

    def all(self) -> list[Todo]:
        return list(self._todos.values())
